Deprecate support for `byzantium` and older EVM versions

cameel opened the pull request:
Resolves #14475.
~Depends on #14590.~ Merged.

cameel committed:
Issue a deprecation warning for byzantium and older EVM versions

cameel commented on libsolidity/interface/StandardCompiler.cpp:
```diff
@@ -787,6 +787,12 @@ std::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompiler:
 		std::optional<langutil::EVMVersion> version = langutil::EVMVersion::fromString(settings["evmVersion"].asString());
 		if (!version)
 			return formatFatalError(Error::Type::JSONError, "Invalid EVM version requested.");
+		if (version <= EVMVersion::constantinople())
+			ret.errors.append(formatError(
+				Error::Type::Warning,
+				"general",
+				"Support for constantinople and older EVM versions is deprecated and will be removed in the future."
+			));
```
I decided to put it in the interface (rather than compiler/yul stack) to avoid having to deal with filtering out the warning in soltest. The EVM version is a CLI argument so the warning would be present or missing in every test based on that.

The consequence is that it has no error code and that it won't appear when the use is not through CLI or Standard JSON (which I guess never happens for a normal user). The upside is that I don't have to track what uses the EVM version and instead I issue the warning at the point where the option gets its value.

cameel commented on Changelog.md:
```diff
@@ -6,6 +6,7 @@ Language Features:
 
 Compiler Features:
  * Commandline Interface: Use proper severity and coloring also for error messages produced outside of the compilation pipeline.
+ * EVM: Deprecate support for "Constantinople" and older EVM versions.
```
Maybe we should deprecate `petersburg` as well? https://github.com/ethereum/solidity/issues/14475#issuecomment-1750986407

ekpyron replied:
In my mind, the plan was actually to deprecate versions **older** than constantinople, **not including** constantinople itself... not sure why I made it ``<=`` in the title of https://github.com/ethereum/solidity/issues/14475, sorry about that.

EDIT: I changed the title there now - the main goal so far is to get rid of ``homestead`` really - we're fine with still supporting constantinople and petersburg for a while.

cameel replied:
Changed.

cameel commented on Changelog.md:
```diff
@@ -6,6 +6,7 @@ Language Features:
 
 Compiler Features:
  * Commandline Interface: Use proper severity and coloring also for error messages produced outside of the compilation pipeline.
+ * EVM: Deprecate support for "Constantinople" and older EVM versions.
```
Should I mention the deprecation somewhere in the docs?

ekpyron replied:
https://docs.soliditylang.org/en/develop/using-the-compiler.html#target-options would seem like an optimal place, wouldn't it?

cameel replied:
Done.

ekpyron commented on test/cmdlineTests/evm_version_constantinople/err:
```diff
@@ -0,0 +1,4 @@
+Warning: Support for constantinople and older EVM versions is deprecated and will be removed in the future.
+
```
Without looking through the PR completely so far, it's a bit disconcerting that the newline here is apparently contingent on whether it's a release build or not (or where is https://app.circleci.com/pipelines/github/ethereum/solidity/31397/workflows/2727648f-a34a-4ea7-a27e-72eb0b4c2470/jobs/1398946 coming from?)

cameel replied:
It's a leftover from the pre-release warning being stripped. Looks like the warning has an empty line after it and we're not stripping it. It went unnoticed probably because stripping of leading whitespace from the output used to remove it anyway. Now, with another warning before it, it no longer happens.

I tweaked the regexes to remove the empty line.

cameel committed:
docs: Fix broken style in the description of shanghai EVM features

cameel committed:
cmdlineTests.sh: Remove the empty line after the pre-release warning as well

- Until now it was being removed along with the leading/trailing whitespace that we strip so it did not matter. Now the EVM deprecation warning can get before this warning, preserving the whitespace and making CLI tests fail.

cameel commented:
Done.

ekpyron commented on test/cmdlineTests.sh:
```diff
@@ -240,9 +240,9 @@ EOF
         rm "$stdout_path.bak"
     else
-        sed -i.bak -e '/^Warning: This is a pre-release compiler version, please do not use it in production./d' "$stderr_path"
+        sed -i.bak -e '/^Warning: This is a pre-release compiler version, please do not use it in production./,+1d' "$stderr_path"
```
What does this do :-)?

cameel replied:
Magic :)

It deletes the match and one line after it.

ekpyron replied:
And why is that something we *want* to do?

cameel replied:
It fixes the problem you pointed out above: https://github.com/ethereum/solidity/pull/14591#discussion_r1350392159

Errors and warnings gathered during compilation are always printed separated with a single empty line. If we remove the warning, but not the empty line, that line will be there for pre-release versions but not for release versions. This makes the stripped CLI output inconsistent between `force_release` jobs and other jobs.

ekpyron replied:
Ah, yeah, sorry, I didn't re-read that after being resolved :-).

nikola-matic commented on docs/using-the-compiler.rst:
```diff
@@ -147,14 +147,14 @@ Target Options
 at each version. Backward compatibility is not guaranteed between each version.
 
-- ``homestead``
+- ``homestead`` (*support deprecated*)
```
Support can't be deprecated, so I'd just change this to
```suggestion
- ``homestead`` (*deprecated*)
```

cameel replied:
I intentionally worded it that way. It's not the EVM version that we're deprecating - we have no control over that and I have no idea if these old EVMs are considered deprecated or not. AFAIK there's no deprecation procedure for hard forks. They just naturally fall out of use. This fuzziness about how relevant they still are is exactly what has been holding us back from doing this deprecation earlier.

What we're doing is deprecating compiler's support for these EVM versions.

I agree that the wording is a bit awkward but I did not find a better way to word it concisely.

nikola-matic replied:
Eh, alright, fair enough. Merge away.

nikola-matic commented on Changelog.md:
```diff
@@ -7,6 +7,7 @@ Language Features:
 Compiler Features:
  * Commandline Interface: Add ``--no-import-callback`` option that prevents the compiler from loading source files not given explicitly on the CLI or in Standard JSON input.
  * Commandline Interface: Use proper severity and coloring also for error messages produced outside of the compilation pipeline.
+ * EVM: Deprecate support for "homestead", "tangerineWhistle", "spuriousDragon" and "byzantium" EVM versions.
```
```suggestion
 * EVM: Deprecate "homestead", "tangerineWhistle", "spuriousDragon" and "byzantium" EVM versions.
```

matheusaaguiar reviewed (approved):
LGTM

nikola-matic reviewed (commented):
Implementation looks good, but the wording of "support deprecated" is odd; deprecation is discouragement of use, so if you deprecate something, you're no longer providing support, and are advising against its use.

nikola-matic reviewed (changes requested):
Slapping a change request since it's already approved, and I want to discuss the wording (as explained in the above comments) before it gets merged.

nikola-matic commented:
Support discontinued?

cameel commented:
But it's not discontinued yet. Just deprecated :)
